sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.scala

```diff
@@ -0,0 +1,432 @@
+      }
+    }
+  }
+
+  /**
+   * Return true if there exists more data in the next batch. If exists, prepare the next batch
+   * by copying from ORC VectorizedRowBatch columns to Spark ColumnarBatch columns.
+   */
+  private def nextBatch(): Boolean = {
+    if (rowsReturned >= totalRowCount) {
+      return false
+    }
+
+    rows.nextBatch(batch)
+    val batchSize = batch.size
+    if (batchSize == 0) {
+      return false
+    }
+    rowsReturned += batchSize
+    columnarBatch.reset()
+    columnarBatch.setNumRows(batchSize)
+
+    var i = 0
+    while (i < requiredSchema.length) {
+      val field = requiredSchema(i)
+      val toColumn = columnVectors(i)
+
+      if (requestedColIds(i) < 0) {
+        toColumn.appendNulls(batchSize)
+      } else {
+        val fromColumn = batch.cols(requestedColIds(i))
+
+        if (fromColumn.isRepeating) {
+          if (fromColumn.isNull(0)) {
+            toColumn.appendNulls(batchSize)
+          } else {
+            field.dataType match {
+              case BooleanType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0) == 1
+                toColumn.appendBooleans(batchSize, data)
+
+              case ByteType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toByte
+                toColumn.appendBytes(batchSize, data)
+              case ShortType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toShort
+                toColumn.appendShorts(batchSize, data)
+              case IntegerType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toInt
+                toColumn.appendInts(batchSize, data)
+              case LongType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0)
+                toColumn.appendLongs(batchSize, data)
+
+              case DateType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toInt
+                toColumn.appendInts(batchSize, data)
+
+              case TimestampType =>
+                val data = fromColumn.asInstanceOf[TimestampColumnVector]
+                toColumn.appendLongs(batchSize, fromTimestampColumnVector(data, 0))
+
+              case FloatType =>
+                val data = fromColumn.asInstanceOf[DoubleColumnVector].vector(0).toFloat
+                toColumn.appendFloats(batchSize, data)
+              case DoubleType =>
+                val data = fromColumn.asInstanceOf[DoubleColumnVector].vector(0)
+                toColumn.appendDoubles(batchSize, data)
+
+              case StringType =>
+                val data = fromColumn.asInstanceOf[BytesColumnVector]
+                var index = 0
+                while (index < batchSize) {
+                  toColumn.appendByteArray(data.vector(0), data.start(0), data.length(0))
+                  index += 1
+                }
+              case BinaryType =>
```

Comment:
Can we merge two cases for `StringType` and `BinaryType` into one case?

Reply:
Yep.